src/dataflow/sink/kafka.rs

```diff
@@ -62,12 +83,15 @@ pub fn kafka<G>(
                     for _ in 0..diff.abs() {
                         let record = BaseRecord::<&Vec<u8>, _>::to(&connector.topic).payload(&buf);
                         if let Err((e, _)) = producer.send(record) {
+                            // TODO(rkhaitan): can we yield and retry sending
+                            // if the error is specifically a Kafka backpressure
+                            // error?
                             error!("unable to produce in {}: {}", name, e);
                         }
                     }
-                    producer.flush(None);
                 }
-            })
+            });
+            producer.flush(std::time::Duration::from_millis(10));
```

Comment:
So I think this roughly adds 10ms of blocking for no reason! We don't particularly care about flushing in any particular time frame, and the threaded producer is already calling poll on our behalf periodically, so the things we've produced _will_ get flushed eventually. 

The only time we actually need to flush is when the sink is getting dropped. (And, arguably, we don't even need to flush then, because I think if you drop a sink there are no guarantees about how much data you'll see.) So I think we should just entirely remove the call to flush, assuming that works.

Reply:
I think the only scenario where we need to flush on a time frame / drop might be tests where the volume of data is very small? testing it out rn compile times are :(

Reply:
The Rust docs are very unclear, but generally without *some* call to flush there is no guarantee that the data make it anywhere other than in a local cache. There are delivery notifications we could look at, but as we don't currently do that either, it's probably best to keep this around until we do start checking asynchronous completions.

Reply:
> The Rust docs are very unclear, but generally without some call to flush there is no guarantee that the data make it anywhere other than in a local cache.

That's not how librdkafka works, actually. As long as you are periodically calling poll, there is no need to call `flush`. The implementation of flush logically looks like this:

```rust
while producer.outstanding_messages > 0 {
    producer.poll()
}
```

The whole point of the threaded producer is to run this loop:

```rust
while !dropped {
    producer.poll()
}
```

So there is no benefit to calling `poll` during standard operation. 

The only problem arises when you drop the threaded producer; the poll loop goes away, potentially before all futures have been delivered. That's the moment at which you need to call poll.

The test failure here actually surfaces an important invariant. If you create a sink on a _closed_ source (like a view of VALUES), then we should publish the entire source to the sink before termination. What's happening in that failed testdrive test is that we queue up delivery futures for all of the data, and then timely realizes "oh hey! nothing left to do" and drops the producer, including the poll loop, before all of the futures have been delivered.

The best way to fix _that_ at the moment is probably something like this:

```
struct FlushingProducer {
    inner: rdkafka::ThreadedProducer,
}

impl Drop for FlushingProducer {
    fn drop(&mut self) {
        self.inner.flush(None);
    }
}
```

Reply:
Right, this
```rust
while producer.outstanding_messages > 0 {
    producer.poll()
}
```
is the logic that is otherwise absent from our code. I was responding to the proposal 

> So I think we should just entirely remove the call to flush, assuming that works.

Reply:
Right, sorry, I'd forgotten about the invariant that we promise to send the whole source if it is closed at the time that you create the sink.